### ISSUE TITLE
Placeholders look broken when block is narrow, e.g. when used in a grid

### ISSUE THREAD

**Issue by noisysocks** [WordPress/gutenberg#64288]
Blocks with placeholder states do not scale to very narrow containers. Example here of an Image placeholder in a narrow grid layout:

![image](https://github.com/user-attachments/assets/601314a0-e46a-42ac-8b88-6f41693bdab7)

There's a separate issue for updating just the Image block has been updated (#54867), but the problem is present for every block that uses the `Placeholder` component. The same solution can be applied:

<img width="1460" alt="i2 Generic placeholders" src="https://github.com/user-attachments/assets/9d14f93f-f100-4a6e-862c-ae3d63267832">

Key here is leaning into a few principles:

* The placeholder content (title, text, buttons) are available for large and medium contexts.
* In small contexts, we rely on the _block toolbar_ for having a single primary CTA. This affords a text label that will not be cropped.
* In that same small context, either the gray unselected placeholder material is used, or the entire spot is a plus button that opens the quick inserter (as it is in empty columns or groups, for example).

[Figma](https://www.figma.com/design/f4rGFqRCB2x0v4yo28dSUN/Placeholders?node-id=37-3096&t=XlmTnnJA8tsGMFZK-1).

**Comment by talldan** [comment]
Similar issue here with some design suggestions - [Difficulties for interacting with the image block when is in a circle form with a small size](https://github.com/WordPress/gutenberg/issues/63254#top)

edit: there are also some mockups here for site logo to deal with the same kind of issue - https://github.com/WordPress/gutenberg/issues/63382#issuecomment-2222699236

**Comment by jasmussen** [comment]
This is a problem with every block that uses a `Placeholder` component, and it has been on our radar for a while now, however in practice it has been hard to simplify it, in part due to back-compat support for components using this in the wild, in part because people easily forget the narrow contexts in which blocks can be placed, and ask with positive intents: why are you removing the title?

This context is nevertheless a good forcing function. Rather than hold firm on legacy elements, let's ensure the component can evolve to support these narrow contexts. Our best bet, most likely, is the built-in responsiveness already there. When the component is as small as here, a `is-small` class should already be present on the component. We can use this to reduce the block to its most important piece—opening the media library. Dan well refers also to the Site Logo (thank for the reference!) which in fact was designed from the start knowing the limitations we had with narrow contexts. 

I've been working on some placeholders [here](https://www.figma.com/design/f4rGFqRCB2x0v4yo28dSUN/Placeholders?node-id=0-1&t=tPJQLAHGRgOzYOgr-1), and a fresh take might apply the exact same treatment for Image as we have in place for the Site Logo:

<img width="1532" alt="i2 Image" src="https://github.com/user-attachments/assets/1b175248-0a24-45e9-84fe-5734562e3425">

The block toolbar will always serve as additional affordance for interacting with the block, and as an identifier for complex patterns that uses many small placeholders together:

<img width="488" alt="i2 Site Logo" src="https://github.com/user-attachments/assets/611f9f64-57a4-493a-8c5f-d2e4142953e6">

**Comment by jasmussen** [comment]
Penny for your thoughts, @WordPress/gutenberg-design

**Comment by jasmussen** [comment]
Also relevant:

* #54867
* #63327

**Comment by jameskoster** [comment]
In the spirit of ideas being cheap; did we ever consider leaving the placeholder abstract, and putting the tools in the block toolbar?

<img width="729" alt="placeholder" src="https://github.com/user-attachments/assets/3a8f6ea9-7f7b-44ca-aab9-6725bf26cd02">

**Comment by jasmussen** [comment]
Works for me, although I would not keep add "Upload" to the toolbar—it's already available through the media library itself, and through the dropzone.

**Comment by jasmussen** [comment]
Given the thumbs up, perhaps this one is ready to update with final mockups and then move to dev? If you all agree, happy to do the honors.

**Comment by jasmussen** [comment]
Created this mockup:

<img width="1763" alt="i3 Image" src="https://github.com/user-attachments/assets/544ea0b2-3f9b-451c-8be0-68b41738a63f">

Realizing that this issue is for _all_ placeholders, I'm going to spend a bit more time on generic patterns that apply across, and instead go ahead and update 54867 for now.

**Comment by vipul0425** [comment]
Hi @jasmussen Can we have a single option in the toolbar that, when clicked, will display all three options available in our large screen placeholder, similar to the replace option? Example:

![image](https://github.com/user-attachments/assets/ecf44657-dc42-4095-a213-cc97b4495695)
![image](https://github.com/user-attachments/assets/ddb7dfcd-a169-4892-a566-416e0884627d)

**Comment by jasmussen** [comment]
I just commented in https://github.com/WordPress/gutenberg/issues/54867#issuecomment-2273257968 about that. Nice work! I'd be here for the dropdown. I've pinged folks for a review on the other issue, which would be a better target for your PR, since  it fixes mainly the _Image_ block, whereas this issue is for the generic issue that affects _all_ blocks.

**Comment by jasmussen** [comment]
Coming back to this one, the issue dedicated just to the Image block has been updated (#54867). This issue aims to solve the same issue, but in a generic way across all placeholders. Here's a sketch at how that might work:

<img width="1460" alt="i2 Generic placeholders" src="https://github.com/user-attachments/assets/9d14f93f-f100-4a6e-862c-ae3d63267832">

Key here is leaning into a few principles:

* The placeholder content (title, text, buttons) are available for large and medium contexts.
* In small contexts, we rely on the _block toolbar_ for having a single primary CTA. This affords a text label that will not be cropped.
* In that same small context, either the gray unselected placeholder material is used, or the entire spot is a plus button that opens the quick inserter (as it is in empty columns or groups, for example).

Let me know your thoughts! @WordPress/gutenberg-design 

[Figma](https://www.figma.com/design/f4rGFqRCB2x0v4yo28dSUN/Placeholders?node-id=37-3096&t=XlmTnnJA8tsGMFZK-1).

**Comment by jasmussen** [comment]
Based on the feedback on this thread, I've gone ahead and updated the issue and shuffled the labels! This still needs a dev, although #54867 is underway (and please help that PR land as well), that will only fix the issue for the Image block. It needs to be solved for the general component as well.

**Comment by t-hamano** [comment]
Update: #54867 added block controls to the Image block to always add media, and in narrow containers, the Image block placeholder always shows with the illustration. A similar approach could be applied to other blocks.

![image](https://github.com/user-attachments/assets/c4c347b3-67e3-48ba-8735-08e15daee2c3)

However, there is an issue with slight flashing when inserting blocks in narrow containers:

https://github.com/user-attachments/assets/066563f5-ce9b-498a-b9e7-a727c148e72e

We've tried to solve this problem in #54867, but haven't found an ideal solution yet.

I'd love your feedback on how to fix this.

**Comment by ndiego** [comment]
@t-hamano @jasmussen now that https://github.com/WordPress/gutenberg/issues/54867 has been merged, can this issue be removed from the 6.7 project board, or punted to 6.8. I know not all of the placeholders have been addressed.

**Comment by jasmussen** [comment]
Yes, can definitely be removed. Thanks!

**Comment by t-hamano** [comment]
I agree. This issue could be considered an enhancement and it can be punted since it contains so many tasks.

**Comment by jasmussen** [comment]
Noting I opened https://github.com/WordPress/gutenberg/issues/66563, which is related to this.

**Comment by vipul0425** [comment]
Hi @t-hamano,

I’ve raised [PR #70840](https://github.com/WordPress/gutenberg/pull/70840) to address the issue for the `Query Loop` block. Looking forward to your review 🙇

Additionally, I noticed that when no patterns are available, users are left without a way to start from a template in the small view state, unlike the placeholder view, which offers a secondary "start blank" option. I’m exploring a solution to conditionally show the "Choose variation" option in the toolbar when no patterns exist. Let me know if this is something we can consider.

https://github.com/user-attachments/assets/7acaf703-9b73-4035-bd2d-6610a4da37eb

cc: @jasmussen
